source/pkcs11.c

```diff
@@ -398,3 +427,316 @@ void aws_pkcs11_lib_release(struct aws_pkcs11_lib *pkcs11_lib) {
         aws_ref_count_release(&pkcs11_lib->ref_count);
     }
 }
+
+/**
+ * Find the slot that meets all criteria:
+ * - has a token
+ * - if match_slot_id is non-null, then slot IDs must match
+ * - if match_token_label is non-null, then labels must match
+ * The function fails unless it finds exactly one slot meeting all criteria.
+ */
+int aws_pkcs11_lib_find_slot_with_token(
+    struct aws_pkcs11_lib *pkcs11_lib,
+    const unsigned long *match_slot_id,
+    const struct aws_string *match_token_label,
+    unsigned long *out_slot_id) {
+
+    CK_SLOT_ID *slot_id_array = NULL;
+    CK_SLOT_ID *candidate = NULL;
+    CK_TOKEN_INFO info;
+    AWS_ZERO_STRUCT(info);
+    bool success = false;
+
+    /* query number of slots with tokens */
+    CK_ULONG num_slots = 0;
+    CK_RV rv = pkcs11_lib->function_list->C_GetSlotList(CK_TRUE /*tokenPresent*/, NULL /*pSlotList*/, &num_slots);
+    if (rv != CKR_OK) {
+        s_raise_ck_error(pkcs11_lib, "C_GetSlotList", rv);
+        goto except;
+    }
+
+    if (num_slots == 0) {
+        AWS_LOGF_ERROR(AWS_LS_IO_PKCS11, "id=%p: No PKCS#11 tokens present in any slot.", (void *)pkcs11_lib);
+        aws_raise_error(AWS_IO_PKCS11_ERROR);
+        goto except;
+    }
+
+    AWS_LOGF_TRACE(
+        AWS_LS_IO_PKCS11, "id=%p: Found %lu slots with tokens. Picking one...", (void *)pkcs11_lib, num_slots);
+
+    /* allocate space for slot IDs */
+    slot_id_array = aws_mem_calloc(pkcs11_lib->allocator, num_slots, sizeof(CK_SLOT_ID));
+
+    /* query all slot IDs */
+    rv = pkcs11_lib->function_list->C_GetSlotList(CK_TRUE /*tokenPresent*/, slot_id_array, &num_slots);
+    if (rv != CKR_OK) {
+        s_raise_ck_error(pkcs11_lib, "C_GetSlotList", rv);
+        goto except;
+    }
+
+    for (size_t i = 0; i < num_slots; ++i) {
+        CK_SLOT_ID slot_id_i = slot_id_array[i];
+
+        /* if specific slot_id requested, and this isn't it, then skip */
+        if ((match_slot_id != NULL) && (*match_slot_id != slot_id_i)) {
+            AWS_LOGF_TRACE(
```

Comment:
sticking with TRACE. this prints once per slot and this [link](http://www.pkiglobe.org/pkcs11_terminology.html) says:
> In case of HSMs, there could be hundreds or more slots are available> 

there's a DEBUG at the end of the function, that tells which slot we ended up picking